src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp

```diff
@@ -0,0 +1,485 @@
+    this->opNavFilterMsgBuffer.timeTag = this->timeTag;
+    eigenMatrixXd2CArray( 1e3*this->state, this->opNavFilterMsgBuffer.state);
+    eigenMatrixXd2CArray( 1e3*this->xBar, this->opNavFilterMsgBuffer.stateError);
+    eigenMatrixXd2CArray(1e6*this->covar, this->opNavFilterMsgBuffer.covar);
+
+    this->navTransOutMsg.write(&this->navTransOutMsgBuffer, this->moduleID, CurrentSimNanos);
+    this->opNavFilterMsg.write(&this->opNavFilterMsgBuffer, this->moduleID, CurrentSimNanos);
+}
+
+/*! This method performs the time update for the flyby OD kalman filter.
+ It propagates the sigma points forward in time and then gets the current
+ covariance and state estimates.
+ @return void
+ @param configData The configuration data associated with the OD filter
+ @param updateTime The time that we need to fix the filter to (seconds)
+ */
+void FlybyODuKF::timeUpdate(double updateTime)
+{
+    Eigen::VectorXd propagatedSigmaPoint;
+    Eigen::MatrixXd A(this->state.size(), 3*this->state.size());
+
+    this->dt = updateTime - this->timeTag;
+    std::array<double, 2> time = {0, this->dt};
+
+    /*! - Copy over the current state estimate into the 0th Sigma point and propagate by dt*/
+    this->sigmaPoints.col(0) = propagator(time, this->state, this->dt);
+    /*! - Scale that Sigma point by the appopriate scaling factor (Wm[0])*/
+    this->xBar = this->wM(0)*this->sigmaPoints.col(0);
+
+    /*! - For each Sigma point, apply sBar-based error, propagate forward, and scale by Wm just like 0th.
+     Note that we perform +/- sigma points simultaneously in loop to save loop values.*/
+    for (size_t i = 1; i<this->state.size() + 1; i++)
+    {
+        /*! - Adding covariance columns from sigma points*/
+        this->sigmaPoints.col(i) = propagator(time, this->state + this->eta * this->sBar.col(i-1), this->dt);
+        /*! - Subtracting covariance columns from sigma points*/
+        this->sigmaPoints.col( i + this->state.size()) =
+                propagator(time, this->state - this->eta * this->sBar.col(i-1), this->dt);
+    }
+
+    /*! - Compute xbar according to Eq (19)*/
+    for (size_t i = 1; i<this->numberSigmaPoints; i++)
+    {
+        this->xBar += this->wM(i)*this->sigmaPoints.col(i);
+    }
+
+    /*! - Assemble the A matrix for QR decomposition as seen in equation 20 in the reference document*/
+    for (size_t i = 1; i < this->numberSigmaPoints; i++)
+    {
+        A.col(i-1) = sqrt(this->wC(i))*(this->sigmaPoints.col(i) - this->xBar);
+    }
+
+    A.block(0, this->numberSigmaPoints-1, this->state.size(), this->state.size()) =
+            this->sQnoise;
+
+    /*! - QR decomposition (only R is of interest) of the A matrix provides the new sBar matrix*/
+    this->sBar = FlybyODuKF::qrDecompositionJustR(A);
+
+    /*! - Shift the sBar matrix over by the xBar vector using the appropriate weight
+     like in equation 21 in design document.*/
+    Eigen::VectorXd xError;
```

Comment:
:large_blue_circle: You could declare/assign this in one line if you like.

Reply:
Done